src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,231 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final int NOT_FOUND_INDEX = -1;
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+
+        if (head == null && tail == null) {
+            head = newNode;
+            tail = newNode;
+            head.prev = null;
+            tail.next = null;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+            tail.next = null;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, false);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            addToBegin(newNode);
+            size++;
+        } else if (index == size) {
+            add(value);
+        } else {
+            addToMiddle(newNode, index);
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el: list) {
+            add(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> newNode = null;
+        if (index >= 0 && index < size) {
+            if (firstHalf(index)) {
+                int indexFirstHalf = index + 1;
+                newNode = head;
+                newNode = firstHalfFindNode(newNode, indexFirstHalf);
+            } else {
+                newNode = tail;
+                int indexFromEnd = size - index - 1;
+                newNode = secondHalfFindNode(newNode, indexFromEnd);
+            }
+            return newNode.item;
+        } else {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        if (node != null) {
+            T item = node.item;
+            node.item = value;
+            return item;
+        } else {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
```

Comment:
instead of this, just put this guarding method inside of a getNode. It shouldn't return null, it either returns an Object or exception is thrown